src/operator/contrib/adamw-inl.h

```diff
@@ -211,6 +192,314 @@ struct AdamWUpdate {
+  return all_inferred;
+}
+
+
+template<typename T>
+class Adam_type_identity {
+ public:
+  using type = T;
+};
+
+
+template<typename T>
+class Adam_single_precision {
+ public:
+  using type = float;
+};
+
+template<typename DType, typename MPDType>
+struct MultiAdamKernelParam {
+  static const int N = 50;
+  int count;
+  size_t max_size;
+  size_t sizes[N];
+  DType* weights[N];
+  DType* grad_data[N];
+  MPDType* mean_data[N];
+  MPDType* var_data[N];
+  MPDType* weights32[N];
+  DType* out_data[N];
+  MPDType clip_gradient;
+  MPDType beta1;
+  MPDType beta2;
+  MPDType etas[N];
+  MPDType lrs[N];
+  MPDType wds[N];
+  MPDType epsilon;
+};
+
+template<typename MPDType, bool has_mixed_precision>
+struct MultiMPAdamWKernel {
+  template<typename DType>
+  MSHADOW_XINLINE static void Map(int i, const MultiAdamKernelParam<DType, MPDType>& param,
+                                  const OpReqType req, const float rescale_grad){
+    for (int index = 0; index < param.count; ++index) {
+      if ((size_t)i < param.sizes[index]) {
+        MPDType w = has_mixed_precision ? param.weights32[index][i]:
+                                          MPDType(param.weights[index][i]);
+        MPDType scaled_grad = static_cast<MPDType>(rescale_grad)*
+                              static_cast<MPDType>(param.grad_data[index][i]);
+
+        if (param.clip_gradient >= 0.0f)
+          scaled_grad = mshadow_op::clip::Map(scaled_grad, param.clip_gradient);
+
+        const auto mean = param.beta1 * (param.mean_data[index][i]- scaled_grad) + scaled_grad;
+        const auto adj = mshadow_op::square::Map(scaled_grad);
+        const auto var = param.beta2 * (param.var_data[index][i] - adj) + adj;
+
+        param.mean_data[index][i] = mean;
+        param.var_data[index][i] = var;
+        w = w - param.etas[index] * (param.lrs[index] *
+            mean / (mshadow_op::square_root::Map(var) + param.epsilon)
+            + param.wds[index] * w);
+        if (has_mixed_precision)
+          param.weights32[index][i] = w;
+
+        KERNEL_ASSIGN(param.out_data[index][i], req, w);
+      }
+    }
+  }
+};
+
+template<typename xpu,
+         typename DType,
+         typename MPDType,
+         typename ParamType = MultiAdamWParam,
+         int input_stride = 4>
+void FillMultiAdamKernelParam(const nnvm::NodeAttrs& attrs,
+                              const OpContext &ctx,
+                              const std::vector<TBlob> &inputs,
+                              const std::vector<TBlob> &outputs,
+                              MultiAdamKernelParam<DType, MPDType> *pParam) {
+  const ParamType& p = nnvm::get<ParamType>(attrs.parsed);
+  mxnet_op::Stream<xpu>* s = ctx.get_stream<xpu>();
+  pParam->clip_gradient = p.clip_gradient;
+  pParam->beta1 = p.beta1;
+  pParam->beta2 = p.beta2;
+
+  pParam->epsilon = p.epsilon;
+
+  pParam->count = p.num_weights;
+  pParam->max_size = 0;
+  const bool isSame = std::is_same<DType, MPDType>::value;
```

Comment:
`constexpr` since it is stronger than `const`.

Reply:
I will make this change.